### PR TITLE
Remove allocations and improve Header Quality Values parsing performance

### PR DESCRIPTION
**Pull request by Yves57**
See #700 
Do not allocate List if not necessary in `HttpHeaderParser.TryParseValues()`
Use the `IList` properties to remove a `foreach`.
Create a new parser (faster perfs and less memory use) for Quality values
Performances (to parse 10 millions `StringWithQualityHeaderValue` values) :
- Before: 3.2s
- After: 1.4s

Memory use (for 10000 parsings):
- Before: 
  before
- After:
  after

### CODE REVIEW

File: src/Microsoft.Net.Http.Headers/HeaderUtilities.cs [hunk 1]
```diff
@@ -60,8 +60,8 @@ internal static void SetQuality(IList<NameValueHeaderValue> parameters, double?
                 // Note that the RFC requires decimal '.' regardless of the culture. I.e. using ',' as decimal
                 // separator is considered invalid (even if the current culture would allow it).
                 double qualityValue;
-                if (double.TryParse(qualityParameter.Value, NumberStyles.AllowDecimalPoint,
-                    NumberFormatInfo.InvariantInfo, out qualityValue))
+                int length;
```
Comment (Tratcher) [inline src/Microsoft.Net.Http.Headers/HeaderUtilities.cs]: unused parameter

File: src/Microsoft.Net.Http.Headers/HeaderUtilities.cs [hunk 2]
```diff
@@ -208,6 +208,100 @@ public static bool TryParseInt64(string value, out long result)
             return long.TryParse(value, NumberStyles.None, NumberFormatInfo.InvariantInfo, out result);
         }
 
+        internal static bool TryParseQualityDouble(string input, int startIndex, out double quality, out int length)
+        {
+            // Fast RFC2616 3.9 Quality value parser (and without memory allocation)
```
Comment (Tratcher) [inline src/Microsoft.Net.Http.Headers/HeaderUtilities.cs]: This comment belongs above the method.

File: src/Microsoft.Net.Http.Headers/HeaderUtilities.cs (hunk 1 above)
Comment (Yves57) [inline src/Microsoft.Net.Http.Headers/HeaderUtilities.cs]: I agree, but what do you want me to do?

File: src/Microsoft.Net.Http.Headers/HttpHeaderParser.cs [hunk 3]
```diff
@@ -61,13 +61,14 @@ protected virtual bool TryParseValues(IList<string> values, bool strict, out ILi
...
             // can ignore the value.
             parsedValues = null;
-            var results = new List<T>();
+            List<T> results = null;
             if (values == null)
             {
                 return false;
             }
-            foreach (var value in values)
+            for (var i = 0; i < values.Count; i++)
```
Comment (davidfowl) [inline src/Microsoft.Net.Http.Headers/HttpHeaderParser.cs]: 👍

File: src/Microsoft.Net.Http.Headers/HeaderUtilities.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.Net.Http.Headers/HeaderUtilities.cs]: Nevermind, I thought it was also unused by the other caller

**Comment by Tratcher** [comment]
@muratg can we get this assigned for review?

**Comment by Tratcher** [comment]
*ping* @muratg

**Comment by muratg** [comment]
We'll get back to this after 1.1.0 is out.

File: src/Microsoft.Net.Http.Headers/HeaderUtilities.cs (hunk 1 above)
Comment (cesarbs) [inline src/Microsoft.Net.Http.Headers/HeaderUtilities.cs]: With C# 7 we can now do `out var` here.

**Pull request closed without merging**